lib/src/hidden_ranges/hidden_ranges.dart

```diff
@@ -160,6 +160,67 @@ class HiddenRanges {
     );
   }
 
+  /// Translates the [position] in the full text to the position in
+  /// the visible text.
+  ///
+  /// If [position] is hidden, returns the last visible position before it.
+  int cutPosition(int position) {
+    if (ranges.isEmpty || position < ranges.first.start) {
+      return position;
+    }
+
+    if (position >= ranges.last.end) {
+      return position - hiddenCharactersBeforeRanges.last;
+    }
+
+    int lowerRange = 0;
+    int upperRange = ranges.length - 1;
+
+    // Linear interpolation search.
+    while (upperRange > lowerRange) {
+      // Full characters at the lower ange start and at the upper range end.
+      final lowerChar = ranges[lowerRange].start;
+      final upperChar = ranges[upperRange].end;
+
+      final rangeIndex = lowerRange +
+          ((position - lowerChar) /
+                  (upperChar - lowerChar) *
+                  (upperRange - lowerRange))
+              .floor();
+
+      if (rangeIndex < lowerRange) {
+        return position - hiddenCharactersBeforeRanges[lowerRange];
+      }
+      if (rangeIndex > upperRange) {
+        return position - hiddenCharactersBeforeRanges[upperRange + 1];
+      }
+
+      final range = ranges[rangeIndex];
+
+      switch (range.compareToPosition(position)) {
+        case 0:
```

Comment:
MAy be will be more readable, if numbers will be changed on "equals", "upper", and "lower"

Reply:
`-1`, `0`, and `1` are common for `sign`-like methods.